solr/solrj/src/java/org/apache/solr/client/solrj/io/ClassificationEvaluation.java

```diff
@@ -36,28 +36,26 @@ public void count(int actual, int predicted) {
     }
   }
 
-  @SuppressWarnings({"unchecked"})
-  public void putToMap(@SuppressWarnings({"rawtypes"})Map map) {
+  public void putToMap(Map<? super String, ? super Long> map) {
     map.put("truePositive_i",truePositive);
     map.put("trueNegative_i",trueNegative);
     map.put("falsePositive_i",falsePositive);
     map.put("falseNegative_i",falseNegative);
   }
 
-  @SuppressWarnings({"rawtypes"})
-  public Map toMap() {
-    HashMap map = new HashMap();
+  public Map<String, Long> toMap() {
```

Comment:
`putToMap` has `Map<? super String, ? super Long>` and I wonder if `toMap` could have `Map<? super String, ? super Long>` return type then to match? Though for either I struggle to see why one might want to use something other than `Map<String, Long>` here i.e. why someone might pass something else to `putToMap` or to override the `toMap` method to return something else.

Reply:
There is some nuance here for why the signatures are different. The mnemonic I've heard that helps me remember is PECS (producer-extends, consumer-super). When we putToMap, we need the caller to know that they have to give us a map that can handle strings and longs. It could be Object,Object, and that's sufficient, the important bit is that somebody won't pass a Map<String,String> where they later will expect the values to be something they're not. When we return from toMap, we explicitly declare what we have, and while this does limit that somebody can't assign it to a Map<Object,Object> anymore, putting bounded wildcards in the return type doesn't let them do that either.